ONSdigital/eq-questionnaire-runner
# Publish confirmation email request to Pub/Sub

### What is the context of this PR?
Publishes confirmation email fulfilment requests to Pub/Sub. I have made some naming changing to the existing IR fulfilment request to make the code a bit more explicitly. Open to suggestions to make it clearer.

Also fixes a minor bug where submitting an email from the thank you page when the limit has already been reached return a 404 page instead of reloading the page.

### How to review 
Ensure you can request a confirmation email using Pub/Sub and that the fulfilment request payload is as expected.

For dev, the fulfilment request is outputted to logs. To use GCP Pub/Sub, you need to deploy the infrastructure.

---

1. **__INITIAL SET UP__**

---

**TESTING LOCALLY**: (Using launcher and runner locally with GCP Pub/Sub)
	
 - Build a GCP environment using Concourse/Terraform with the infrastructure repo pointed at https://github.com/ONSdigital/eq-terraform-gcp/pull/23 (this will create a Pub/Sub topic called `eq-submission-confirmation-topic`)
 - Authenticate with gcloud SDK using:
	- `gcloud auth login`
	- `gcloud auth application-default login`
	- `gcloud config set project <your-project-id>`
 - Set the following env vars:
	- `EQ_PUBLISHER_BACKEND: pubsub`
	- `EQ_SUBMISSION_CONFIRMATION_TOPIC_ID: eq-submission-confirmation-topic`


**TESTING IN GCP**: (Using deployed launcher and runner with GCP Pub/Sub)
  - Build a GCP environment using Concourse/Terraform
    -  Using the following branch when deploying the infrastructure https://github.com/ONSdigital/eq-terraform-gcp/pull/23
    - Set the runner branch and image tag. Since we don't push to GCR for runner, you will need to build and push your own image to GCR to deploy to Concourse. Alternatively you can use the image tag I have already pushed: `queue-confirmation-email`
		
	
---

2. __**TEST MESSAGE PUBLICATIONS**__

---
  - Go to Pub/Sub in GCP Console and click the `eq-submission-confirmation-topic` topic.
  - Click `VIEW MESSAGES`, this will prompt you to create a `Subscription` which is needed to view messages.
  - Request a confirmation email using `test_confirmation_email` schema.
  - Click `VIEW MESSAGES` in the GCP UI, select the subscription you just created and click `PULL`. You should now see the confirmation email fulfilment request payload. Test multiple scenarios and ensure the payload matches the specification.


### Checklist
